### PR TITLE
Prevent conflicts with overlapping calls

### LINKED ISSUE #3: Overlapping calls seem to return the first available result
By inspection, it seems like the following scenario will not work:

1. main calls answerRenderer for channel "makeUpper"
2. renderer calls callMain for channel "makeUpper" and data "aaa"
3. renderer calls callMain for channel "makeUpper" and data "bbb"
4. Callback for "makeUpper" of "aaa" returns "AAA"
5. Both renderer calls to callMain receive this data (they are using the same channel)
6. Code thinks makeUpper of "bbb" returned "AAA"

When I implemented something similar (before this module was available), I needed a unique identifier to correlate request/response and avoid this problem: https://github.com/DavidAnson/Lenz/blob/master/ipc.js

Might something like that be necessary here?

### PR DESCRIPTION
**Pull request by javierron**
fixes #3

### CODE REVIEW

File: source/main.js [hunk 1]
```diff
@@ -6,29 +6,38 @@ const {ipcMain: ipc, BrowserWindow} = electron;
 
 ipc.callRenderer = (window, channel, data) => new Promise((resolve, reject) => {
 	const {sendChannel, dataChannel, errorChannel} = util.getRendererResponseChannels(window.id, channel);
+	const id = `${Date.now()}.${Math.random()}`;
```
Comment (sindresorhus) [inline source/main.js]:
```suggestion
	const id = `${Date.now()}-${Math.random()}`;
```

File: source/main.js [hunk 2]
```diff
@@ -6,29 +6,38 @@ const {ipcMain: ipc, BrowserWindow} = electron;
...
 
-	ipc.on(errorChannel, (event, error) => {
+	ipc.on(uniqueErrorChannel, (event, error) => {
 		cleanup();
 		reject(error);
 	});
 
+	const completeData = {
+		uniqueDataChannel,
+		uniqueErrorChannel,
```
Comment (sindresorhus) [inline source/main.js]: Let's call it just `dataChannel` and `errorChannel` in the sent data.

```js
const completeData = {
	dataChannel: uniqueDataChannel,
	errorChannel: uniqueErrorChannel,
```

File: source/util.js [hunk 3]
```diff
@@ -1,13 +1,13 @@
 'use strict';
 
 module.exports.getResponseChannels = channel => ({
-	sendChannel: `%better-ipc-send-channel-${channel}`,
-	dataChannel: `%better-ipc-response-data-channel-${channel}`,
-	errorChannel: `%better-ipc-response-error-channel-${channel}`
```
Comment (sindresorhus) [inline source/util.js]: Don't remove this. The prefix here is still useful for debugging.

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "electron-better-ipc",
-	"version": "0.1.1",
+	"version": "0.1.2-rc5",
```
Comment (sindresorhus) [inline package.json]: Don't bump the version. I will do so when appropriate.

File: source/main.js [hunk 5]
```diff
@@ -6,29 +6,38 @@ const {ipcMain: ipc, BrowserWindow} = electron;
...
+		content: data
```
Comment (sindresorhus) [inline source/main.js]:
```suggestion
		userData: data
```

File: source/renderer.js [hunk 6]
```diff
@@ -1,39 +1,50 @@
...
+const electron = window.require('electron');
 const util = require('./util');
 
 const {ipcRenderer: ipc} = electron;
 
 ipc.callMain = (channel, data) => new Promise((resolve, reject) => {
 	const {sendChannel, dataChannel, errorChannel} = util.getResponseChannels(channel);
+	const id = `${Date.now()}.${Math.random()}`;
+	const uniqueDataChannel = `${dataChannel}-${id}`;
+	const uniqueErrorChannel = `${errorChannel}-${id}`;
```
Comment (sindresorhus) [inline source/renderer.js]: This is the same as the one in `main.js`. You could move them to `util.js` as functions that return unique channels.

**Comment by oeddyo** [comment]
any update?

**Comment by oeddyo** [comment]
ping :P

**Comment by sindresorhus** [comment]
I'm willing to accept https://github.com/sindresorhus/electron-better-ipc/pull/9/commits/43bd92013e923a3ac87587a1a414a460386a1ce3, but it should be a separate PR, and it should be documented, including why it's needed. I'm taking it out of this PR.

**Pull request merged**